File: src/p1xxx/p1531.py
from leetcode_prelude import *


# Problem 1531
class Solution:
    def getLengthOfOptimalCompression(self, s: str, k: int) -> int:
        dp = [[9999] * (k + 1) for _ in range(len(s) + 1)]
        dp[0][:] = [0] * (k + 1)

        count2len = [0] * 101
        count2len[1] = 1
        count2len[2:10] = [2] * 8
        count2len[10:100] = [3] * 90
        count2len[100] = 4

        for i in range(len(s) + 1):
            for j in range(0, k + 1):
                length = dp[i][j]

                if j > 0:
                    length = min(length, dp[i - 1][j - 1])

                start = i - 1
                stop = i
                deleted = 0
                ch = s[start]

                length = min(length, dp[start][j - deleted] + 1)

                for start in range(stop - 1, -1, -1):
                    if s[start] != ch:
                        if deleted >= j:
                            break
                        deleted += 1

                    count = stop - start - deleted
                    encoded_len = count2len[count]

                    length = min(length, dp[start][j - deleted] + encoded_len)

                dp[i][j] = length

        return dp[len(s)][k]

File: src/p1xxx/p1531.rs
mod leetcode_prelude;

use std::ffi::FromVecWithNulError;

use leetcode_prelude::*;

fn main() {
    println!("123456");

    use std::hint::black_box;

    println!("456789");
}

struct Solution {}

extern crate rand;

// Problem 1531
impl Solution {
    pub fn get_length_of_optimal_compression(s: String, k: i32) -> i32 {
        let k = k as usize;

        // dp[i][j] is the minimum length when encoding s[0..i] with at most j characters deleted
        // dp size is (s.len() + 1)-by-(k + 1)
        let mut dp: Vec<Vec<i32>> = Vec::with_capacity(s.len() + 1);

        dp.push(vec![0; k + 1]); // zero size input is always zero
        dp.resize(s.len() + 1, vec![i32::MAX; k + 1]);

        assert!(s.len() < dp.len());

        for i in 1..=s.len() {
            for j in 0..=k {
                // encoding s[0..i] with at most j characters deleted
                // s[i - 1] is the last character in the substring

                // case 1: delete s[i - 1]
                if j > 0 {
                    dp[i][j] = dp[i][j].min(dp[i - 1][j - 1]);
                }

                // case 2: keep s[i - 1]
                // try to encode s[..i + 1] backward with at most j characters deleted

                // a better name to represent the end of substring
                let stop = i;
                let mut start = i - 1;
                let mut deleted = 0; // how many deletions done

                // encode it alone, ie. encode(s[..start]) + s[start]
                dp[i][j] = dp[i][j].min(dp[start][j - deleted] + 1);

                let ch = s.bytes().nth(start).unwrap();

                while start > 0 {
                    start -= 1;

                    if s.bytes().nth(start).unwrap() != ch {
                        if deleted >= j {
                            // no deletions are available. Cannot encode s[start..i]
                            break;
                        }

                        deleted += 1;
                    }

                    // how many time ch is repeated along some characters deleted
                    let count = stop - start - deleted;

                    // encoded length of s[start..stop] with some characters deleted
                    let encoded_len = match count {
                        1 => 1,
                        2..=9 => 2,
                        10..=99 => 3,
                        100..=999 => 4,
                        _ => unreachable!(),
                    };
                    // so the total length depends on how many deletions left.
                    dp[i][j] = dp[i][j].min(dp[start][j - deleted] + encoded_len);
                }
            }
        }

        dp[s.len()][k]
    }
}
